src/pathfinding.cpp

```diff
@@ -192,12 +192,19 @@ std::vector<tripoint> map::route( const tripoint &f, const tripoint &t,
     }
     // First, check for a simple straight line on flat ground
     // Except when the line contains a pre-closed tile - we need to do regular pathing then
-    if( f.z == t.z && clear_path( f, t, -1, 2, 2 ) ) {
+    constexpr auto non_normal = PF_SLOW | PF_WALL | PF_VEHICLE | PF_TRAP;
+    if( f.z == t.z ) {
         const auto line_path = line_to( f, t );
-        const std::set<tripoint> sorted_line( line_path.begin(), line_path.end() );
-
-        if( is_disjoint( sorted_line, pre_closed ) ) {
-            return line_path;
+        const auto &pf_cache = get_pathfinding_cache_ref( f.z );
+        // Check all points for any special case (including just hard terrain)
+        if( std::all_of( line_path.begin(), line_path.end(), [&pf_cache]( const tripoint & p ) {
+        return !( pf_cache.special[p.x][p.y] & non_normal );
+        } ) ) {
```

Comment:
This condition looks a bit confusing. It would be more readable like this:
```c++
const bool simple_route = std::none_of( line_path.begin(), line_path.end(), [&pf_cache]( const tripoint & p ) {
    return pf_cache.special[p.x][p.y] & non_normal;
} );

if( simple_route ) {
    ...
```